File: java/arcs/core/storage/Driver.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

import kotlin.reflect.KClass

/** Listener for changes to the [Data] managed by a [Driver]. */
typealias DriverReceiver<Data> = suspend (data: Data, version: Int) -> Unit

/**
 * Interface that all drivers must support.
 *
 * Note the threading of a version number here; each model provided by the [Driver] to the [ActiveStore]
 * (using a receiver registered with [registerReceiver]) is paired with a version, as is each model
 * sent from the [ActiveStore] to the driver (using [send]).
 *
 * This threading is used to track whether driver state has changed while the [ActiveStore] is processing
 * a particular model. [send] should always fail if the version isn't exactly `1` greater than the
 * current internal version.
 *
 * Note that the [Driver] is *not* a thread-safe interface! Calls to [registerReceiver] and [send]
 * should take place from the same thread that creates the Driver.
 */
interface Driver<Data : Any> {
  /** Key identifying the [Driver]. */
  val storageKey: StorageKey
  /** Kotlin class for provided Data generic, i.e. class representing what's being stored. */
  val dataClass: KClass<Data>
  /**
   * Returns a token that represents the current state of the data.
   *
   * This can be provided to [registerReceiver], and will impact what data is delivered on
   * initialization (only "new" data should be delivered, though note that this can be satisfied
   * by sending a model for merging rather than by remembering a set of ops).
   */
  val token: String?

  /** Registers a listener for [Data]. */
  suspend fun registerReceiver(token: String? = null, receiver: DriverReceiver<Data>)

  /**
   * Sends data to the [Driver] for storage.
   *
   * If the data is stored successfully, this method will call any registered [DriverReceiver]
   * with the stored [Data] and updated version, before returning True to the caller.
   *
   * If the data is not stored successfully, this method will return False. This in turn implies
   * that the caller should wait for its [DriverReceiver] to be called before attempting to
   * call [send] again.
   **/
  suspend fun send(data: Data, version: Int): Boolean

  /** Closes the driver and releases any held resources. */
  suspend fun close() = Unit

  /** Creates a copy of this [Driver] that is registered to the same storage. */
  suspend fun clone(): Driver<Data>
}

File: src/runtime/storage/drivers/driver.ts
/**
 * @license
 * Copyright (c) 2019 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {StorageKey} from '../storage-key.js';

export enum Exists {ShouldExist, ShouldCreate, MayExist}
export type ReceiveMethod<T> = (model: T, version: number) => void;

// Interface that drivers must support.
//
// Note the threading of a version number here; each model provided
// by the driver to the Store (using the receiver) is paired with a version,
// as is each model sent from the Store to the driver (using Driver.send()).
//
// This threading is used to track whether driver state has changed while
// the Store is processing a particular model. send() should always fail
// if the version isn't exactly 1 greater than the current internal version.
export abstract class Driver<Data> {
    storageKey: StorageKey;
    exists: Exists;
    constructor(storageKey: StorageKey, exists: Exists) {
      this.storageKey = storageKey;
      this.exists = exists;
    }
    abstract registerReceiver(receiver: ReceiveMethod<Data>, token?: string): void;
    abstract async send(model: Data, version: number): Promise<boolean>;

    // Return a token that represents the current state of the data.
    // This can be provided to registerReceiver, and will impact what
    // data is delivered on initialization (only "new" data should be
    // delivered, though note that this can be satisfied by sending
    // a model for merging rather than by remembering a set of ops)
    abstract getToken(): string | null;

    // these methods only available to Backing Stores and will
    // be removed once entity mutation is performed on CRDTs
    // tslint:disable-next-line: no-any
    abstract async write(key: StorageKey, value: any): Promise<void>;
    // tslint:disable-next-line: no-any
    abstract async read(key: StorageKey): Promise<any>;
  }
